lib/classes/Variables.js

```diff
@@ -962,33 +962,53 @@ class Variables {
     );
   }
 
-  warnIfNotFound(variableString, valueToPopulate) {
-    if (
+  handleUnresolved(variableString, valueToPopulate) {
+    const valueEmpty =
       valueToPopulate === null ||
       typeof valueToPopulate === 'undefined' ||
       (typeof valueToPopulate === 'object' &&
         !Array.isArray(valueToPopulate) &&
-        !Object.keys(valueToPopulate).length)
-    ) {
-      let varType;
-      if (variableString.match(this.envRefSyntax)) {
-        varType = 'environment variable';
-      } else if (variableString.match(this.optRefSyntax)) {
-        varType = 'option';
-      } else if (variableString.match(this.selfRefSyntax)) {
-        varType = 'service attribute';
-      } else if (variableString.match(this.fileRefSyntax)) {
-        varType = 'file';
-      } else if (variableString.match(this.ssmRefSyntax)) {
-        varType = 'SSM parameter';
-      }
-      if (varType) {
-        logWarning(
-          `A valid ${varType} to satisfy the declaration '${variableString}' could not be found.`
-        );
-      }
+        !Object.keys(valueToPopulate).length);
+
+    if (!valueEmpty) {
+      return;
+    }
+
+    let varType;
+
+    if (variableString.match(this.envRefSyntax)) {
+      varType = 'environment variable';
+    } else if (variableString.match(this.optRefSyntax)) {
+      varType = 'option';
+    } else if (variableString.match(this.selfRefSyntax)) {
+      varType = 'service attribute';
+    } else if (variableString.match(this.fileRefSyntax)) {
+      varType = 'file';
+    } else if (variableString.match(this.ssmRefSyntax)) {
+      varType = 'SSM parameter';
+    }
+
+    if (!varType) {
+      return;
     }
-    return valueToPopulate;
+
+    if (this.service.unresolvedVariablesNotificationMode === 'error') {
+      throw new this.serverless.classes.Error(
+        `A valid ${varType} to satisfy the declaration '${variableString}' could not be found.`
+      );
+    }
+
+    // only log the deprecation method if this isn't being *explicitly* set to 'warn' (and is not 'error')
+    if (this.service.unresolvedVariablesNotificationMode !== 'warn') {
```

Comment:
It can be simplified to `if (this.service.unresolvedVariablesNotificationMode)` (by schema, if it exists, it can only be `warn`)

Reply:
I felt this was clearer than `!this.service.unresolvedVariablesNotificationMode`, but up to you - could also do `=== undefined` 🤷

Reply:
Cleanest would be `!this.service.unresolvedVariablesNotificationMode`.

Technically if anyone is setting this mode to any value, no deprecation should be shown.